spec/index.bs

```diff
@@ -587,107 +587,126 @@ the user behind an `ArrayBuffer`, thus avoiding any data copies.
+    ::  Source of a copy operation. Preserves the contents.
+        Allowed by {{GPUBufferUsage/COPY_SRC}} or {{GPUTextureUsage/COPY_SRC}}.
+    : <dfn>copy-destination</dfn>
+    ::  Destination of a copy operation.
+        Allowed by {{GPUBufferUsage/COPY_DST}} or {{GPUTextureUsage/COPY_DST}}.
+    : <dfn>input</dfn>
+    :: Buffer with input data for draw or dispatch calls. Preserves the contents.
+        Allowed by {{GPUBufferUsage/INDEX}}, {{GPUBufferUsage/VERTEX}}, or {{GPUBufferUsage/INDIRECT}}.
+    : <dfn>uniform</dfn>
+    ::  Resource bindings that are uniformly accessed by the shaders. Preserves the contents.
+        Allowed by {{GPUBufferUsage/UNIFORM}} or {{GPUTextureUsage/SAMPLED}}.
+    : <dfn>storage</dfn>
+    ::  Writable storage resource binding.
+        Allowed by {{GPUBufferUsage/STORAGE}} or {{GPUTextureUsage/STORAGE}}.
+    : <dfn>storage-read</dfn>
+    ::  Read-only storage resource bindings. Preserves the contents.
+        Allowed by {{GPUBufferUsage/STORAGE}} or {{GPUTextureUsage/STORAGE}}.
+    : <dfn>attachment</dfn>
+    :: Texture used as an output attachment in a render pass.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+    : <dfn>attachment-read</dfn>
+    ::  Texture used as a read-only attachment in a render pass. Preserves the contents.
+        Allowed by {{GPUTextureUsage/OUTPUT_ATTACHMENT}}.
+</dl>
 
-Issue(gpuweb/gpuweb#296): Consider merging all read-only usages.
+Issue: as copy operations are specified manually
+(like, "source buffer has to be different from the destination buffer"),
+should we remove the "copy-*" variants,
+or try to refactor the copy operations spec to match?
 
 Textures may consist of separate [=mipmap levels=] and [=array layers=],
 which can be used differently at any given time.
-Each such <dfn dfn>subresource</dfn> is uniquely identified by a
+Each such <dfn dfn>texture subresource</dfn> is uniquely identified by a
 [=texture=], [=mipmap level=], and
 (for {{GPUTextureDimension/2d}} textures only) [=array layer=].
 
-The **main usage rule** is that any [=subresource=]
-at any given time can only be in either:
-  - a combination of [=read-only usage=]s
-  - a single [=mutating usage=]
+We define <dfn dfn>subresource</dfn> to be either a whole buffer, or a [=texture subresource=].
 
-Enforcing this rule allows the API to limit when data races can occur
-when working with memory. That property makes applications written against
+Some [=internal usage=]s are compatible with others.
+That means an implementation can consider multiple usages active at the same time.
+There is a limited set of <dfn dfn>allowed usage set</dfn>, where each usage is compatible with the other:
+  - {[=copy-source=] | [=input=] | [=uniform=] | [=storage-read=]} - in any combination
+  - [=copy-destination=] - singular usage
+  - {[=storage=] | [=storage-read=]} - in any combination
+  - [=attachment=] - singular usage
+  - {[=attachment-read=] | [=uniform=]} - in any combination
```

Comment:
What about storage read?

Reply:
I'm not sure if this is actually allowed :/  I guess, depending on https://github.com/KhronosGroup/Vulkan-Docs/issues/1345 we'll either allow storage-read here, or scrap the whole RODS out of WebGPU (or make it an extension?). So there is no harm in specifying here that storage-read is allowed, until we know more.